src/virt.rs

```diff
@@ -314,58 +314,150 @@ impl VirtContext {
+        rd: Register,
+        rs1: Register,
+        imm: usize,
+        multiplier: usize,
+        length: usize,
+    ) {
+        let address = self.get(rs1) + imm * multiplier;
+        let offset = address - device.start_addr;
+        match device.device_interface.read_device(offset) {
+            Ok(mut value) => {
+                let mask = if length < usize::BITS as usize {
+                    (1 << length) - 1
+                } else {
+                    usize::MAX
+                };
+                value &= mask;
+
+                self.set(rd, value);
+                self.pc += 2;
             }
-            Instr::Sd { rs1, rs2, imm } => {
-                let address = utils::calculate_addr(self.get(rs1), *imm);
-                let offset = address - device.start_addr;
-                match device.device_interface.write_device(offset, self.get(rs2)) {
-                    Ok(()) => self.pc += 4,
-                    Err(err) => panic!("Error writing {}: {}", device.name, err),
-                }
+            Err(err) => panic!("Error reading {}: {}", device.name, err),
+        }
+    }
+    fn handle_c_store(
+        &mut self,
+        device: &VirtDevice,
+        rs1: Register,
+        rs2: Register,
+        imm: usize,
+        multiplier: usize,
+        length: usize,
+    ) {
+        let address = self.get(rs1) + imm * multiplier;
+        let offset = address - device.start_addr;
+        let mut value = self.get(rs2);
+
+        let mask = if length < usize::BITS as usize {
+            (1 << length) - 1
+        } else {
+            usize::MAX
+        };
+
+        if value > mask {
+            log::warn!(
+                "Value {} exceeds allowed length {}. Trimming to fit.",
+                value,
+                length
+            );
+            value &= mask;
+        }
+        match device.device_interface.write_device(offset, value) {
+            Ok(()) => {
+                self.pc += 2;
             }
-            Instr::CSd { rs1, rs2, imm } => {
-                let address = self.get(rs1) + imm * 8;
-                let offset = address - device.start_addr;
-                match device.device_interface.write_device(offset, self.get(rs2)) {
-                    Ok(()) => self.pc += 2,
-                    Err(err) => panic!("Error writing {}: {}", device.name, err),
+            Err(err) => panic!("Error writing {}: {}", device.name, err),
+        }
+    }
+
+    fn handle_load(
+        &mut self,
+        device: &VirtDevice,
+        rd: Register,
+        rs1: Register,
+        imm: isize,
+        length: usize,
+    ) {
+        let address = utils::calculate_addr(self.get(rs1), imm);
+        let offset = address - device.start_addr;
+        match device.device_interface.read_device(offset) {
+            Ok(mut value) => {
+                // Sign-extend to 32 bits
+                if length < 32 && value < (1 << 31) {
+                    value = (value as i32) as usize;
```

Comment:
Out of curiosity why is there a 32 bits extend here? Is is mandated by some aspect of the spec or related to our own implementation?

Reply:
From the spec (v2.2, p.19):

> LH loads a 16-bit value from memory, then sign-extends to 32-bits before storing in rd ... LB is defined analogously for 8-bit values

I took it as LH and LB should sign-extend the value. For stores no extension is mentioned.